### ISSUE TITLE
Why is CodeFixer run thrice to apply the actual fix instead of being cached?

### ISSUE THREAD

**Issue by ghost** [dotnet/roslyn#32460]
**Version Used**: 
2.6.3
**Steps to Reproduce**:

```csharp
class StaticInheritanceAnalyzer 
{ 
    internal static ConcurrentDictionary<Location, DiagnosticData> DiagnosticDataDict =
            new ConcurrentDictionary<Location, DiagnosticData>();
} // gets poplated by all the necesary data to ensure static inheritance.
// also: can't use context.RegisterCompilationStartAction to clear DiagnosticDataDict. because it wipes data arbitrarily and unreliably.`
```

CodeFixer:

```csharp
        private ConcurrentDictionary<Location, int> cacheReleaseCounter = new ConcurrentDictionary<Location, int>();

        private async Task<Document> FixInheritanceAsync(Document document, Location location,
            CancellationToken cancellationToken)
        {
            DiagnosticData diagnosticData;
            int counter = cacheReleaseCounter.GetOrAdd(location, 0);
            switch (counter)
            {
                case 0:
                case 1:
                    StaticInheritanceAnalyzer.DiagnosticDataDict.TryGetValue(location, out diagnosticData);
                    break;
                case 2:
                    StaticInheritanceAnalyzer.DiagnosticDataDict.TryRemove(location, out diagnosticData);
                    break;
                default:
                    throw new Exception(counter.ToString());
            }
            cacheReleaseCounter.TryUpdate(location, counter == 2 ? 0 : counter + 1, counter);
            // counter == 2 => fix gets applied, analyzer will finally produce a new Solution
            [... actual code ...]
        }
```

1. Cache some DiagnosticData
2. Try to release Data on first CodeFix run
3. Run into the Issue that first Run is only used for the Preview
4. Wonder what the next two runs do
5. Be happy that it at least needs 3 runs consistently ( for what I've encountered so far)

**Expected Behavior**:

cached Data can be released by first call / CodeFixer needs to be run only once

**Actual Behavior**:

Codefixer has to be run thrice before Fix is applied

**Comment by mavasani** [comment]
Couple of things:
1. Storing any data statically on an analyzer or code fix is not recommended. That is almost certain to cause memory leaks.
2. Tagging @heejaechang who might know why we execute the code fix multiple times for preview - note that this is purely a _performance_ question. Any assumption or functional dependency on code fixer being invoked just once on a given snapshot is incorrect and should never be made.

**Comment by Draethorian** [comment]
Well Hmm, I need some Values: 
* The location of baseMember
* The locations of the baseMembers which need to be inherited
* The fixed Attribute list (which isn't a big deal to calculate after figuring out which need to be inherited [AttrUsage Inherited = true], so please don't blame for doing that  with the Analyzer, could probably make a task for it). They're a list of SyntaxReferences but Locations should also work.

So is there any way I could convert Locations to string and back?

Concerning the memory leak: Instead of calling DiagnosticDataDict.Clear when context.RegisterCompilationStartAction is called. I call Clear when the Fix is applied the third time.. So it is just delayed.

**Comment by CyrusNajmabadi** [comment]
> So is there any way I could convert Locations to string and back?

You can just use '.AdditionalLocations' on the diagnostic.  Lots of analyzers/fixers work this way.

**Comment by jmarolf** [comment]
> I'm eager to get to know more complex Design Patterns. that would be amazing 💪 🎉

In general we hope what most analyzers/codefixes can be stateless and deterministic.  This allows us to use parallelism better as well as moving your stuff out of the Visual Studio process so less memory is consumed. Calling [EnableConcurrentExecution](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,198) essentially tells us that you meet these requirements.

As you've seen we offer some stateful APIs for analyzers so you can stash information such as symbol definitions.  Attempting to store state in a different way is generally a way to get hard-to-diagnose bugs.

Because your analyzer may live in a different process then your codefix there are only a few ways the two can communicate.  Anything that you place on the `Diagnostic` type we will serialize and move across processes for you.  In-general if more information than this needs to be passed we typically recommend recomputing it on the codefix side.

**Comment by Draethorian** [comment]
> You can just use '.AdditionalLocations' on the diagnostic. Lots of analyzers/fixers work this way.

😞 really? damn I feel retarded^^ I know that I've seen that more than a half dozen times but aways thought, meh, not what I need... Bummer... AHH now I get it. I thought that was used internally for mulitple symbolLocations within a block or sth... Alright that's why I didn't use it. Thanks ;-)


> In general we hope what most analyzers/codefixes can be stateless and deterministic. This allows us to use parallelism better as well as moving your stuff out of the Visual Studio process so less memory is consumed. Calling [EnableConcurrentExecution](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,198) essentially tells us that you meet these requirements.
> 
> As you've seen we offer some stateful APIs for analyzers so you can stash information such as symbol definitions. Attempting to store state in a different way is generally a way to get hard-to-diagnose bugs.
> 
> Because your analyzer may live in a different process then your codefix there are only a few ways the two can communicate. Anything that you place on the `Diagnostic` type we will serialize and move across processes for you. In-general if more information than this needs to be passed we typically recommend recomputing it on the codefix side.

Ah, thanks for the Insight. Sounds amazing what you're doing ;-) I was just about tinker a custom CodeFixAllProvider^^ Good you're here ;-) Now I'm even more interested how Roslyn works under the Hood. Is there any "How we did it: Roslyn's Software Architecture Special" Youtube Video out there? would be awesome. 😄 Again thanks for your help guys. I'm doing some quick fixes now ;-)

**Comment by jmarolf** [comment]
> "How we did it: Roslyn's Software Architecture Special" Youtube Video out there?

Josh Varity did a [set of youtube videos a while back](https://www.youtube.com/watch?v=wXXHd8gYqVg&list=PLxk7xaZWBdUT23QfaQTCJDG6Q1xx6uHdG) that is mostly about getting started. Unfortunately I can't point you at something more in depth than that at the moment.

**Comment by Draethorian** [comment]
Yeah, that's what got me started.  I thought more about how those mentioned
Processes are intergrated into VS. But maybe I find something when binging
it properly ;-)

Am Mi., 16. Jan. 2019 um 21:12 Uhr schrieb Jonathon Marolf <
notifications@github.com>:

**Comment by Draethorian** [comment]
You know what the biggest bummer of all is: I figured static Inheritance is utter crap for using it dynamically, which was my damn f-ing stupid intention 😅. I really didn't figure that I need an object to cast up the inheritance chain. I'm so freaking, fucking retarded. I wasted the last 2 months on that crap and I simply missed the obvious. And simply because I couldn't accept that it was forbidden and bad practice, and didn't know why. (here goes my Career at Microsoft) So geesh, here you got it. I'm fucking retarded. But hey, maybe I can let people know that (Static Inheritance isn't that usable^^).

Geez, that would be the most embarassing Blog and Stackoverflow Entry ever. HA! Let's do it 😄
